kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/KubernetesCrudDispatcher.java

```diff
@@ -75,4 +148,248 @@ public MockResponse handleGet(String path) {
+      for (AttributeSet item : items) {
+        for (Map.Entry<AttributeSet, Set<WatchEventsListener>> entry: watchEventListenerMapping.entrySet()) {
+          if (item.matches(entry.getKey())) {
+            for (WatchEventsListener listener : entry.getValue()) {
+              listener.sendWebSocketResponse(map.get(item), "DELETED");
+            }
+          }
+        }
+        map.remove(item);
+      }
+      response.setResponseCode(200);
+    } else {
+      response.setResponseCode(404);
+    }
+    return response;
+  }
+
+  public MockResponse handleWatch(RecordedRequest request) {
+    MockResponse mockResponse = new MockResponse();
+    String path = request.getPath();
+    String resourceName = fetchResourceNameFromWatchRequestPath(path);
+    AttributeSet query = attributeExtractor.extract(path);
+    if (resourceName != null) {
+      query = query.add(new Attribute("name", resourceName));
+    }
+    WatchEventsListener watchEventListener = new WatchEventsListener(query, UUID.randomUUID());
+    if (watchEventListenerMapping.containsKey(query)) {
+      watchEventListenerMapping.get(query).add(watchEventListener);
+    } else {
+      watchEventListenerMapping.put(query, new HashSet<WatchEventsListener>() {{ add(watchEventListener); }});
+    }
+    mockResponse.setSocketPolicy(SocketPolicy.KEEP_OPEN);
+    return mockResponse.withWebSocketUpgrade(watchEventListener);
+  }
+
+  boolean detectWatchMode(String path) {
+    String queryString = null;
+    try {
+      queryString = new URI(path).getQuery();
+    } catch (URISyntaxException e) {
+      LOGGER.debug("incorrect URI string: " + path);
+      return false;
+    }
+    return queryString == null || queryString.isEmpty()? false: queryString.contains("watch=true");
+  }
+
+  String fetchResourceNameFromWatchRequestPath(String path) {
+    String queryString = null;
+    try {
+      queryString = new URI(path).getQuery();
+    } catch (URISyntaxException e) {
+      LOGGER.debug("Incorrect URI string: " + path);
+      return null;
+    }
+
+    if (queryString == null || queryString.isEmpty()) {
+      return null;
+    }
+
+    String name = "";
+    for (String q: queryString.split("&")) {
+      if (q.contains("fieldSelector") && q.contains("metadata.name")) {
+        String s[] = q.split("=");
+        name = s[s.length - 1];
+      }
+    }
+    return name.isEmpty()? null: name;
+  }
+
+  private String doGet(String path) {
+    List<String> items = new ArrayList<>();
+    AttributeSet query = attributeExtractor.extract(path);
+    for (Map.Entry<AttributeSet, String> entry : map.entrySet()) {
+      if (entry.getKey().matches(query)) {
+        items.add(entry.getValue());
+      }
+    }
+
+    if (items.isEmpty()) {
+      return null;
+    } else if (items.size() == 1) {
+      return items.get(0);
+    } else {
+      return responseComposer.compose(items);
+    }
+  }
+
+  private class WatchEventsListener extends WebSocketListener {
```

Comment:
When declaring inner classes it's' recommended that you make them static, else you'll be creating the class each time the parent class is instantiated.

Anyway, this file is already 400 lines long, why not move this class to a separate file to make things clearer and more readable?